### PR TITLE
[Resource] Third party S2S Consent/Permission Enumeration

### PR DESCRIPTION
**Pull request by 00Kai0**
**Description**

add `az provider permission list` for api: https://docs.microsoft.com/en-us/rest/api/resources/providers/provider-permissions
add arg '--consent-to-permissions' for `az provider register`, api: https://docs.microsoft.com/en-us/rest/api/resources/providers/register#providerconsentdefinition

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Resource

File: src/azure-cli/azure/cli/command_modules/resource/_help.py [hunk 1]
```diff
@@ -1846,6 +1846,16 @@
         az provider list --query [?namespace=='Microsoft.Network'].resourceTypes[].resourceType
 """
 
+helps['provider permission'] = """
+type: group
+short-summary: Get provider permissions.
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/resource/_help.py:1851]:
```suggestion
short-summary: Manage permissions for a provider.
```

File: src/azure-cli/azure/cli/command_modules/resource/_help.py [hunk 2]
```diff
@@ -1846,6 +1846,16 @@
...
+"""
+
+helps['provider permission list'] = """
+type: command
+short-summary: Get permissions from providers.
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/resource/_help.py:1856]:
```suggestion
short-summary: List permissions from a provider.
```

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_provider.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from azure.cli.testsdk import ScenarioTest
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_provider.py:6]: No need to add this test in a separate file as you can use existing test file `test_resource.py`.

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_provider.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+class ProviderTests(ScenarioTest):
+
+    def test_register(self):
+
+        self.kwargs = {
+            'namespace': "Microsoft.PolicyInsights"
+        }
+
+        self.cmd('provider unregister -n {namespace}')
+        self.cmd('provider register -n {namespace} --consent-to-permissions')
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_provider.py:17]: Check output for the commands

File: src/azure-cli/azure/cli/command_modules/resource/_client_factory.py [hunk 5]
```diff
@@ -72,8 +72,8 @@ def cf_resources(cli_ctx, _):
     return _resource_client_factory(cli_ctx).resources
 
 
-def cf_providers(cli_ctx, _):
-    return _resource_client_factory(cli_ctx).providers
+def cf_providers(cli_ctx, **_):
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/resource/_client_factory.py:75]: What does `**_` mean?

File: src/azure-cli/azure/cli/command_modules/resource/_client_factory.py [hunk 6]
```diff
@@ -72,8 +72,8 @@ def cf_resources(cli_ctx, _):
...
+    return _resource_client_factory(cli_ctx, api_version="2021-04-01").providers
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/resource/_client_factory.py:76]: `2021-04-01` will be the default version for client?

**Comment by jsntcy** [comment]
@00Kai0, please add more information in PR description so that reviewers can know the purpose of this PR.
Reference to write a great description for a PR: https://www.pullrequest.com/blog/writing-a-great-pull-request-description/

**Review by jsntcy — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 7]
```diff
@@ -1019,31 +1020,31 @@ def _get_auth_provider_latest_api_version(cli_ctx):
     return api_version
 
 
-def _update_provider(cli_ctx, namespace, registering, wait, mg_id=None, accept_terms=None):
+def _update_provider(cli_ctx, namespace, registering, wait, properties=None, mg_id=None, accept_terms=None):
     import time
     target_state = 'Registered' if registering else 'Unregistered'
-    rcf = _resource_client_factory(cli_ctx)
+    client = cf_providers(cli_ctx, None, api_version='2021-04-01')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:1026]: May I ask why you do not upgrade the api-version for `ResourceType.MGMT_RESOURCE_RESOURCES`, but specify a fixed version `2021-04-01` here?
Actually, If the current api-version does not support new features, we will generally bump the api-version

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 8]
```diff
@@ -2036,8 +2037,12 @@ def list_resources(cmd, resource_group_name=None,
     return list(resources)
 
 
-def register_provider(cmd, resource_provider_namespace, mg=None, wait=False, accept_terms=None):
-    _update_provider(cmd.cli_ctx, resource_provider_namespace, registering=True, wait=wait, mg_id=mg, accept_terms=accept_terms)
+def register_provider(cmd, resource_provider_namespace, consent_to_permissions=False, mg=None, wait=False, accept_terms=None):
+    properties = None
+    if consent_to_permissions:
+        from azure.mgmt.resource.resources.v2021_04_01.models import ProviderRegistrationRequest, ProviderConsentDefinition
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:2043]: It is not recommended to directly import the model from package under the specific version. If you plan to bump api-version, then you can use `cmd.get_models` to get those models

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 9]
```diff
@@ -2049,6 +2054,11 @@ def list_provider_operations(cmd):
     return auth_client.provider_operations_metadata.list()
 
 
+def list_provider_permissions(cmd, resource_provider_namespace):
+    client = cf_providers(cmd.cli_ctx, None, api_version='2021-04-01')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:2058]: Same as above

**Review by zhoxing-ms — APPROVED** [review]
@00Kai0 Because this PR requires this sprint to be released, so it can be merged as a temporary solution. Please remember to bump api-version to implement this feature

**Pull request merged**